sockeye/lexical_constraints.py

```diff
@@ -24,6 +24,200 @@
+                 batch_size: int,
+                 beam_size: int,
+                 avoid_list: Optional[List[RawConstraintList]] = None,
+                 global_avoid_trie: Optional[AvoidTrie] = None) -> None:
+
+        self.global_avoid_states = []  # type: List[AvoidState]
+        self.local_avoid_states = []  # type: List[AvoidState]
+
+        # Store the global trie for each hypothesis
+        if global_avoid_trie is not None:
+            self.global_avoid_states = [AvoidState(global_avoid_trie)] * batch_size * beam_size
+
+        # Store the sentence-level tries for each item in their portions of the beam
+        if avoid_list is not None:
+            for raw_phrases in avoid_list:
+                self.local_avoid_states += [AvoidState(AvoidTrie(raw_phrases))] * beam_size
+
+    def reorder(self, indices: mx.nd.array) -> None:
+        """
+        Reorders the avoid list according to the selected row indices.
+        This can produce duplicates, but this is fixed if state changes occur in consume().
+
+        :param indices: An mx.nd.array containing indices of hypotheses to select.
+        """
+        if self.global_avoid_states:
+            self.global_avoid_states = [self.global_avoid_states[x] for x in indices.asnumpy()]
+
+        if self.local_avoid_states:
+            self.local_avoid_states = [self.local_avoid_states[x] for x in indices.asnumpy()]
+
+    def consume(self, word_ids: mx.nd.array) -> None:
+        """
+        Consumes a word for each trie, updating respective states.
+
+        :param word_ids: The set of word IDs.
+        """
+        word_ids = word_ids.asnumpy().tolist()
+        for i, word_id in enumerate(word_ids):
+            if self.global_avoid_states:
+                self.global_avoid_states[i] = self.global_avoid_states[i].consume(word_id)
+            if self.local_avoid_states:
+                self.local_avoid_states[i] = self.local_avoid_states[i].consume(word_id)
+
+    def avoid(self) -> Tuple[List[int], List[int]]:
+        """
+        Assembles a list of per-hypothesis words to avoid. The indices are (x, y) pairs into the scores
+        array, which has dimensions (beam_size, target_vocab_size). These values are then used by the caller
+        to set these items to np.inf so they won't be selected. Words to be avoided are selected by
+        consulting both the global trie of phrases and the sentence-specific one.
+
+        :return: Two lists of indices: the x coordinates and y coordinates.
+        """
+        to_avoid = set()  # type: Set[Tuple[int, int]]
+        for i, state in enumerate(self.global_avoid_states):
+            for word_id in state.avoid():
+                if word_id > 0:
+                    to_avoid.add((i, word_id))
+        for i, state in enumerate(self.local_avoid_states):
+            for word_id in state.avoid():
+                if word_id > 0:
+                    to_avoid.add((i, word_id))
+
+        x_list = []  # type: List[int]
```

Comment:
this could be simplified to ```list(zip(*to_avoid))```, right? (generating a list of tuples)

Reply:
Ah, yes, thank you! Hated this bit of code.